src/databricks/labs/ucx/hive_metastore/views_migrator.py

```diff
@@ -0,0 +1,122 @@
+        statement = self._parse_view_text()
+        for sql_table in statement.find_all(SqlTable):
+            catalog = self._catalog(sql_table)
+            if catalog != 'hive_metastore':
+                continue
+            table_with_key = Table(catalog, sql_table.db, sql_table.name, "type", "")
+            table = all_tables.get(table_with_key.key)
+            if table is None:
+                raise ValueError(
+                    f"Unknown schema object: {table_with_key.key} in view SQL: {self._view.view_text} of table {self._view.key}"
+                )
+            if table.view_text is None:
+                table_dependencies.add(table)
+            else:
+                view_dependencies.add(table)
+        self._table_dependencies = list(table_dependencies)
+        self._view_dependencies = list(view_dependencies)
+
+    def _parse_view_text(self) -> SqlExpression:
+        try:
+            # below can never happen but avoids a pylint error
+            assert self._view.view_text is not None
+            statements = sqlglot.parse(self._view.view_text)
+            if len(statements) != 1 or statements[0] is None:
+                raise ValueError(f"Could not analyze view SQL: {self._view.view_text} of table {self._view.key}")
+            return statements[0]
+        except ParseError as e:
+            raise ValueError(f"Could not analyze view SQL: {self._view.view_text} of table {self._view.key}") from e
+
+    # duplicated from FromTable._catalog, not sure if it's worth factorizing
+    @staticmethod
+    def _catalog(table):
+        if table.catalog:
+            return table.catalog
+        return 'hive_metastore'
+
+    def __hash__(self):
+        return hash(self._view)
+
+
+class ViewsMigrator:
+
+    def __init__(self, crawler: TablesCrawler):
+        self._crawler = crawler
+        self._result_view_list: list[ViewToMigrate] = []
+        self._result_tables_set: set[Table] = set()
+
+    def sequence(self) -> list[Table]:
+        # sequencing is achieved using a very simple algorithm:
+        # for each view, we register dependencies (extracted from view_text)
+        # then given the remaining set of views to process,
+        # and the growing set of views already processed
+        # we check if each remaining view refers to not yet processed views
+        # if none, then it's safe to add that view to the next batch of views
+        # the complexity for a given set of views v and a dependency depth d looks like Ov^d
+        # this seems enormous but in practice d remains small and v decreases rapidly
+        table_list = self._crawler.snapshot()
+        raw_tables = set(filter(lambda t: t.view_text is None, table_list))
+        raw_views = set(table_list)
+        raw_views.difference_update(raw_tables)
+        table_keys = [table.key for table in table_list]
+        all_tables = dict(zip(table_keys, table_list))
+        views = {ViewToMigrate(view) for view in raw_views}
```

Comment:
```suggestion
     all_tables = dict()
     for table in table_list:
         all_tables[table.key] = table
         if table.view_text is None:
            continue
         views.add(ViewToMigrate(table))
```
This above code can easily be replaced by just these 5 lines. 
Existing code is running 8 iterations over list of tables. 

N(filter - line no 80) + N(set - line no 80) + N(set - line no 81) + N(set difference update - line no 82) + N[list comprehension - line no 83] + N(zip - line no 84) + N(dict - line no 84) + N(set comprehension - line no 85)

While suggestive code is just running a single iteration. 

CC: @nfx

Reply:
To be accurate, operations on sets should not count as full scans since they only operate on hashcodes. But your suggestion is certainly faster anyway.